File: pkg/engine/engine_test.go
package engine

import (
	"testing"

	"warframe-state-graph/pkg/model"
)

// Ash Stealth Build テストケース（Obsidian Prototype/Nodes/ と同一構成）を
// Goエンジンでも再現し、手動トレースで確認した期待結果と一致するか検証する。
func buildTestGraph() *model.Graph {
	g := model.NewGraph()
	g.Nodes["ash-stealth-build"] = &model.Node{
		ID: "ash-stealth-build", Type: model.TypeBuild,
		Contains: []string{"dragon-nikana", "ash-prime", "seeking-shuriken"},
	}
	g.Nodes["dragon-nikana"] = &model.Node{
		ID: "dragon-nikana", Type: model.TypeWeapon, Satisfied: true,
		Contains: []string{"dragon-nikana-riven"},
	}
	g.Nodes["dragon-nikana-riven"] = &model.Node{
		ID: "dragon-nikana-riven", Type: model.TypeRiven, Satisfied: false,
	}
	g.Nodes["ash-prime"] = &model.Node{
		ID: "ash-prime", Type: model.TypeFrame, Satisfied: true,
	}
	g.Nodes["seeking-shuriken"] = &model.Node{
		ID: "seeking-shuriken", Type: model.TypeMod, Satisfied: false,
		Requires: []string{"red-veil-rank-5"},
	}
	g.Nodes["red-veil-rank-5"] = &model.Node{
		ID: "red-veil-rank-5", Type: model.TypeSyndicate, Satisfied: false,
	}
	return g
}

func TestDeriveNextActions_AshStealthBuild(t *testing.T) {
	g := buildTestGraph()
	report := DeriveNextActions(g, "ash-stealth-build")

	wantActionable := map[string]bool{"dragon-nikana-riven": true, "red-veil-rank-5": true}
	wantBlocked := map[string]bool{"seeking-shuriken": true}
	wantSatisfied := map[string]bool{"dragon-nikana": true, "ash-prime": true}

	assertSet(t, "Actionable", report.Actionable, wantActionable)
	assertSet(t, "Blocked", report.Blocked, wantBlocked)
	assertSet(t, "Satisfied", report.Satisfied, wantSatisfied)

	if report.Progress.Done != 2 || report.Progress.Total != 5 {
		t.Errorf("progress = %+v, want {Done:2 Total:5}", report.Progress)
	}
}

func TestResolveState_CyclicRequiresIsBlockedNotInfiniteLoop(t *testing.T) {
	g := model.NewGraph()
	g.Nodes["a"] = &model.Node{ID: "a", Requires: []string{"b"}}
	g.Nodes["b"] = &model.Node{ID: "b", Requires: []string{"a"}}

	state := ResolveState(g, "a", map[string]bool{})
	if state != model.StateBlocked {
		t.Errorf("cyclic requires: got %s, want BLOCKED", state)
	}
}

// Steel Path クエスト連鎖（Saya's Vigil → Chains of Harrow → ... → Natah）を模した
// requiresチェーン。Natahを達成にした際、前提が全て自動で連鎖達成になるか検証する。
func TestCascadeSatisfyRequires_ChainOfPrerequisites(t *testing.T) {
	g := model.NewGraph()
	g.Nodes["natah"] = &model.Node{ID: "natah", Requires: []string{"war-within"}}
	g.Nodes["war-within"] = &model.Node{ID: "war-within", Requires: []string{"second-dream"}}
	g.Nodes["second-dream"] = &model.Node{ID: "second-dream", Requires: []string{"apostasy"}}
	g.Nodes["apostasy"] = &model.Node{ID: "apostasy", Requires: []string{"chains-of-harrow"}}
	g.Nodes["chains-of-harrow"] = &model.Node{ID: "chains-of-harrow", Requires: []string{"saya-vigil"}}
	g.Nodes["saya-vigil"] = &model.Node{ID: "saya-vigil"}

	CascadeSatisfyRequires(g, "natah", map[string]bool{})

	// natah自身はこの関数の対象外（呼び出し元がnode.Satisfied=trueにする想定）。
	// 前提チェーン5つは全て連鎖達成済みになっているはず。
	for _, id := range []string{"war-within", "second-dream", "apostasy", "chains-of-harrow", "saya-vigil"} {
		if !g.Nodes[id].Satisfied {
			t.Errorf("%s: want satisfied=true after cascade, got false", id)
		}
	}
}

func TestCascadeSatisfyRequires_CyclicDoesNotInfiniteLoop(t *testing.T) {
	g := model.NewGraph()
	g.Nodes["a"] = &model.Node{ID: "a", Requires: []string{"b"}}
	g.Nodes["b"] = &model.Node{ID: "b", Requires: []string{"a"}}

	CascadeSatisfyRequires(g, "a", map[string]bool{}) // ハングせず返れば十分
	if !g.Nodes["b"].Satisfied {
		t.Errorf("b: want satisfied=true, got false")
	}
}

// Natahを未達に戻したら、Natahに依存している後工程（Steel Path解放）も連鎖で未達に戻るはず。
// 一方、Natahの前提（The War Within等）は完了実績のままなので触らない。
func TestCascadeUnsatisfyDependents_RevertsDownstreamOnly(t *testing.T) {
	g := model.NewGraph()
	g.Nodes["war-within"] = &model.Node{ID: "war-within", Satisfied: true}
	g.Nodes["natah"] = &model.Node{ID: "natah", Requires: []string{"war-within"}, Satisfied: false}
	g.Nodes["steel-path-junction"] = &model.Node{ID: "steel-path-junction", Requires: []string{"natah"}, Satisfied: true}
	g.Nodes["further-quest"] = &model.Node{ID: "further-quest", Requires: []string{"steel-path-junction"}, Satisfied: true}

	CascadeUnsatisfyDependents(g, "natah", map[string]bool{})

	if !g.Nodes["war-within"].Satisfied {
		t.Errorf("war-within (前提側): want satisfied=true のまま, got false")
	}
	if g.Nodes["steel-path-junction"].Satisfied {
		t.Errorf("steel-path-junction (直接の依存先): want satisfied=false, got true")
	}
	if g.Nodes["further-quest"].Satisfied {
		t.Errorf("further-quest (間接的な依存先): want satisfied=false, got true")
	}
}

func TestCascadeUnsatisfyDependents_CyclicDoesNotInfiniteLoop(t *testing.T) {
	g := model.NewGraph()
	g.Nodes["a"] = &model.Node{ID: "a", Requires: []string{"b"}, Satisfied: true}
	g.Nodes["b"] = &model.Node{ID: "b", Requires: []string{"a"}, Satisfied: true}

	CascadeUnsatisfyDependents(g, "a", map[string]bool{}) // ハングせず返れば十分
	if g.Nodes["b"].Satisfied {
		t.Errorf("b: want satisfied=false, got true")
	}
}

func assertSet(t *testing.T, label string, got []string, want map[string]bool) {
	t.Helper()
	if len(got) != len(want) {
		t.Errorf("%s: got %v, want keys of %v", label, got, want)
		return
	}
	for _, id := range got {
		if !want[id] {
			t.Errorf("%s: unexpected id %q in %v", label, id, got)
		}
	}
}

File: ts/server/engine.test.ts
// Port of pkg/engine/engine_test.go — same 6 cases, same test graphs, same
// expected results, translated to TS naming/idiom. Test-first executable
// spec for the cascade/resolve core.

import { expect, test } from "bun:test";
import { newGraph, type Graph } from "./model.ts";
import {
  cascadeSatisfyRequires,
  cascadeUnsatisfyDependents,
  deriveNextActions,
  resolveState,
} from "./engine.ts";

// Ash Stealth Build test case (same as the Obsidian Prototype/Nodes/ setup),
// reproduced here so a manually-traced expected result can be checked
// against both the Go and TS engines.
function buildTestGraph(): Graph {
  const g = newGraph();
  g.nodes["ash-stealth-build"] = {
    id: "ash-stealth-build",
    name: "",
    type: "Build",
    satisfied: false,
    requires: [],
    contains: ["dragon-nikana", "ash-prime", "seeking-shuriken"],
  };
  g.nodes["dragon-nikana"] = {
    id: "dragon-nikana",
    name: "",
    type: "Weapon",
    satisfied: true,
    requires: [],
    contains: ["dragon-nikana-riven"],
  };
  g.nodes["dragon-nikana-riven"] = {
    id: "dragon-nikana-riven",
    name: "",
    type: "Riven",
    satisfied: false,
    requires: [],
    contains: [],
  };
  g.nodes["ash-prime"] = {
    id: "ash-prime",
    name: "",
    type: "Frame",
    satisfied: true,
    requires: [],
    contains: [],
  };
  g.nodes["seeking-shuriken"] = {
    id: "seeking-shuriken",
    name: "",
    type: "Mod",
    satisfied: false,
    requires: ["red-veil-rank-5"],
    contains: [],
  };
  g.nodes["red-veil-rank-5"] = {
    id: "red-veil-rank-5",
    name: "",
    type: "Syndicate",
    satisfied: false,
    requires: [],
    contains: [],
  };
  return g;
}

function assertSet(got: string[], want: string[]): void {
  expect(new Set(got)).toEqual(new Set(want));
  expect(got.length).toBe(want.length);
}

// Minimal node factory for tests that don't need the full Ash Stealth Build
// setup — mirrors the Go tests' terse &model.Node{ID: ..., Requires: ...}
// literals, filling in the fields Zod's schema requires but the Go struct
// left at zero value.
function node(id: string, opts: { requires?: string[]; satisfied?: boolean } = {}) {
  return {
    id,
    name: "",
    type: "Resource" as const,
    satisfied: opts.satisfied ?? false,
    requires: opts.requires ?? [],
    contains: [],
  };
}

test("deriveNextActions classifies the Ash Stealth Build graph", () => {
  const g = buildTestGraph();
  const report = deriveNextActions(g, "ash-stealth-build");

  assertSet(report.actionable, ["dragon-nikana-riven", "red-veil-rank-5"]);
  assertSet(report.blocked, ["seeking-shuriken"]);
  assertSet(report.satisfied, ["dragon-nikana", "ash-prime"]);

  expect(report.progress.done).toBe(2);
  expect(report.progress.total).toBe(5);
});

test("resolveState treats cyclic requires as BLOCKED, not an infinite loop", () => {
  const g = newGraph();
  g.nodes.a = node("a", { requires: ["b"] });
  g.nodes.b = node("b", { requires: ["a"] });

  expect(resolveState(g, "a")).toBe("BLOCKED");
});

// Steel Path quest chain (Saya's Vigil → Chains of Harrow → ... → Natah)
// modeled as a requires chain. Completing Natah should cascade-complete
// every prerequisite automatically.
test("cascadeSatisfyRequires completes a chain of prerequisites", () => {
  const g = newGraph();
  g.nodes.natah = node("natah", { requires: ["war-within"] });
  g.nodes["war-within"] = node("war-within", { requires: ["second-dream"] });
  g.nodes["second-dream"] = node("second-dream", { requires: ["apostasy"] });
  g.nodes.apostasy = node("apostasy", { requires: ["chains-of-harrow"] });
  g.nodes["chains-of-harrow"] = node("chains-of-harrow", { requires: ["saya-vigil"] });
  g.nodes["saya-vigil"] = node("saya-vigil");

  cascadeSatisfyRequires(g, "natah");

  // natah itself is out of scope for this function (the caller is expected
  // to set node.satisfied = true first); the 5-node prerequisite chain
  // should all be cascade-satisfied.
  for (const id of ["war-within", "second-dream", "apostasy", "chains-of-harrow", "saya-vigil"]) {
    expect(g.nodes[id]?.satisfied).toBe(true);
  }
});

test("cascadeSatisfyRequires on a cycle does not infinite-loop", () => {
  const g = newGraph();
  g.nodes.a = node("a", { requires: ["b"] });
  g.nodes.b = node("b", { requires: ["a"] });

  cascadeSatisfyRequires(g, "a"); // returning without hanging is the test
  expect(g.nodes.b?.satisfied).toBe(true);
});

// Reverting Natah should cascade-revert downstream dependents (Steel Path
// access), while leaving Natah's own prerequisites (The War Within) alone —
// their completion record doesn't get erased.
test("cascadeUnsatisfyDependents reverts downstream only", () => {
  const g = newGraph();
  g.nodes["war-within"] = node("war-within", { satisfied: true });
  g.nodes.natah = node("natah", { requires: ["war-within"], satisfied: false });
  g.nodes["steel-path-junction"] = node("steel-path-junction", { requires: ["natah"], satisfied: true });
  g.nodes["further-quest"] = node("further-quest", { requires: ["steel-path-junction"], satisfied: true });

  cascadeUnsatisfyDependents(g, "natah");

  expect(g.nodes["war-within"]?.satisfied).toBe(true); // prerequisite side: untouched
  expect(g.nodes["steel-path-junction"]?.satisfied).toBe(false); // direct dependent
  expect(g.nodes["further-quest"]?.satisfied).toBe(false); // indirect dependent
});

test("cascadeUnsatisfyDependents on a cycle does not infinite-loop", () => {
  const g = newGraph();
  g.nodes.a = node("a", { requires: ["b"], satisfied: true });
  g.nodes.b = node("b", { requires: ["a"], satisfied: true });

  cascadeUnsatisfyDependents(g, "a"); // returning without hanging is the test
  expect(g.nodes.b?.satisfied).toBe(false);
});
